### PR TITLE
FFWEB-727 documentation 404 page not shown (2nd stage)

### PR DESCRIPTION
**Pull request by omikron-fp**
This is the second stage of the 404 page fix.

You can now define a redirect target inside the page configs. This will be a client-side redirect only.
If a page is not found, there is now an attempt to find a "best-match" page.
It also contains a hard-coded server-side 301 redirect for the old landing page that still shows up in Google results.

### CODE REVIEW

File: src/actions/app.js [hunk 1]
```diff
@@ -24,10 +22,10 @@ export const navigate = (path, params) => (dispatch) => {
 };
 
 const loadPage = (page, subpage, tab) => (dispatch) => {
-    const pageInfo = pageInfoCollection[page];
+    const importInfo = pageImportInfoCollection[page];
 
-    if (pageInfo && (!pageInfo.requiresSubpage || pageInfo.subpages[subpage])) {
-        pageInfo.importTarget();
+    if (importInfo && (!importInfo.requiresSubpage || importInfo.subpages[subpage] && !importInfo.subpages[subpage].hasMoved)) {
```
Comment (KrzysztofKozlowskiOBSS) [inline src/actions/app.js]: quite complicated condition, can we extract some part into a function with a descriptive name?

File: src/data/pageMappings.js [hunk 2]
```diff
@@ -1,23 +1,62 @@
...
 }
+
+function getSubpageSuggestion(subpage) {
+    if (api.pages[subpage])
+        return createSubpageSuggestion(`api`, api.pages[subpage]);
+
+    if (documentation.pages[subpage])
+        return createSubpageSuggestion(`documentation`, documentation.pages[subpage]);
+
+    return undefined;
```
Comment (KrzysztofKozlowskiOBSS) [inline src/data/pageMappings.js]: this return can be skipped I guess

File: src/data/pageMappings.js [hunk 3]
```diff
@@ -1,23 +1,62 @@
...
+function createPageImportInfo(importTargetCall, subpages = undefined) {
     return Object.freeze({
         importTarget: importTargetCall,
         subpages,
         requiresSubpage: !!subpages
     });
 }
+
+function getSubpageSuggestion(subpage) {
+    if (api.pages[subpage])
```
Comment (KrzysztofKozlowskiOBSS) [inline src/data/pageMappings.js]: Personally I don't like skipping curly braces after if's

File: src/util/functors.js [hunk 4]
```diff
@@ -0,0 +1,8 @@
+export function IsDefinedFunctor(value) {
+    return Object.freeze({
+        valueOf: () => value,
+        map: value === undefined || value === null
+            ? () => IsDefinedFunctor()
+            : fn => IsDefinedFunctor(fn(value))
+    })
```
Comment (KrzysztofKozlowskiOBSS) [inline src/util/functors.js]: semicolon

File: src/util/url.js [hunk 5]
```diff
@@ -0,0 +1,11 @@
+export function urlPathToPages(path) {
+    const paths = path.split('/');
+    return {
+        page: paths[1] || `home`,
+        subpage: paths[2]
+    }
```
Comment (KrzysztofKozlowskiOBSS) [inline src/util/url.js]: semicolon

File: src/data/pageMappings.js [hunk 6]
```diff
@@ -1,23 +1,62 @@
...
+        : IsDefinedFunctor(getSubpageSuggestion(subpage));
+}
+
+export function hasMovedModifier(page, newSubpageData) {
+    return createPageInfo(page, newSubpageData, { hasMoved: true });
+}
 
-function createPageInfo(importTargetCall, subpages = undefined) {
+
+function createPageImportInfo(importTargetCall, subpages = undefined) {
```
Comment (KrzysztofKozlowskiOBSS) [inline src/data/pageMappings.js]: isn't `subpages` arg `undefined` by default if it's not provided?

Comment (KrzysztofKozlowskiOBSS) [reply src/data/pageMappings.js]: makes perfect sense

File: src/views/view-404.js [hunk 7]
```diff
@@ -24,20 +26,61 @@ class View404 extends PolymerElement {
...
+        justify-content: center;
+    }
 </style>
 
-<h2>Oops! You hit a 404!</h2>
+
+<h2>Oops! You hit a [[errorCode]]!</h2>
 <div class="imgContainer">
-    <img src="/images/404-cat.jpg">
+    <img src="/images/[[errorCode]]-cat.jpg">
```
Comment (damrath) [inline src/views/view-404.js]: Nice touch :joy_cat:

**Pull request merged**